src/Workspaces/Core/Portable/Workspace/Solution/DocumentState.cs

```diff
@@ -759,5 +761,38 @@ public static DocumentId GetDocumentIdForTree(SyntaxTree tree)
             s_syntaxTreeToIdMap.TryGetValue(tree, out var id);
             return id;
         }
+
+        private static void CheckTree(
+            SyntaxTree newTree,
+            SourceText newText,
+            SyntaxTree oldTree = null,
+            SourceText oldText = null)
+        {
+            // this should be always true
+            if (newTree.Length == newText.Length)
+            {
+                return;
+            }
+
+            var newTreeContent = newTree.GetRoot().ToFullString();
+            var newTextContent = newText.ToString();
+
+            var oldTreeContent = oldTree?.GetRoot().ToFullString();
+            var oldTextContent = oldText?.ToString();
+
+            // we time to time see (incremental) parsing bug where text <-> tree round tripping is broken.
+            // send NFW for those cases
+            FatalError.ReportWithoutCrash(new Exception($"tree and text has different length {newTree.Length} vs {newText.Length}"));
+
+            GC.KeepAlive(newTreeContent);
```

Comment:
Are these all to prevent them from being GC'ed so we still have them in the dump? Make sense, but might be worth commenting.

Reply:
sure.